packages/runtime/container-runtime/src/connectionTelemetry.ts

```diff
@@ -160,16 +162,19 @@ class OpPerfTelemetry {
         }
 
         // Record collab window max size after every 1000th op.
-        if (sequenceNumber % 1000 === 0) {
-            if (this.opSendTimeForLatencyStatisticsForMsnStatistics !== undefined) {
-                this.logger.sendPerformanceEvent({
-                    eventName: "MsnStatistics",
-                    sequenceNumber,
-                    msnDistance: this.deltaManager.lastSequenceNumber - this.deltaManager.minimumSequenceNumber,
-                    duration: message.timestamp - this.opSendTimeForLatencyStatisticsForMsnStatistics,
-                });
-            }
-            this.opSendTimeForLatencyStatisticsForMsnStatistics = message.timestamp;
+        if (this.sequenceNumberForMsnTracking === undefined && sequenceNumber % 1000 === 0) {
+            this.sequenceNumberForMsnTracking = sequenceNumber;
+            this.msnTrackingTimestamp = message.timestamp;
+        }
+        if (this.sequenceNumberForMsnTracking !== undefined &&
+                message.minimumSequenceNumber >= this.sequenceNumberForMsnTracking) {
```

Comment:
>message.minimumSequenceNumber >= this.sequenceNumberForMsnTracking [](http://example.com/codeflow?start=16&length=66)

Out of curiosity, can there ever be a scenario in which message.minimumSequenceNumber == message.sequenceNumber  && sequenceNumber %1000 == 0 ?

Reply:
It should never happen for client ops. Basically, msn tells every client (including future clients who will join) the minimum reference sequence number of future ops. This client (who sent this op) may still sent ops with previous sequence number as a reference sequence number, so MSN can't move yet.

I think the only place where that might be violated is when there are no clients connected to service (i.e. ops are produced by service). last leaveOp or noClient op is the only op I could think of.

I glanced at one file, and it's actually true for both :) I also found summaryAck to have same property (also sent by service). In all these cases clientId == null.

I think it's fine to leave it as is, as we record the truth, even if collab window is empty.